sdk/eventhub/event-hubs/src/eventHubBufferedProducerClient.ts

```diff
@@ -0,0 +1,315 @@
+import { isCredential } from "./util/typeGuards";
+
+/**
+ * Contains the events that were successfully sent to the Event Hub,
+ * and the partition they were assigned to.
+ */
+export interface OnSendEventsSuccessContext {
+  /**
+   * The partition each event was assigned.
+   */
+  partitionId: string;
+  /**
+   * The array of {@link EventData} and/or `AmqpAnnotatedMessage` that were successfully sent to the Event Hub.
+   */
+  events: Array<EventData | AmqpAnnotatedMessage>;
+}
+
+/**
+ * Contains the events that were not successfully sent to the Event Hub,
+ * the partition they were assigned to, and the error that was encountered while sending.
+ */
+export interface OnSendEventsErrorContext {
+  /**
+   * The partition each event was assigned.
+   */
+  partitionId: string;
+  /**
+   * The array of {@link EventData} and/or `AmqpAnnotatedMessage` that were not successfully sent to the Event Hub.
+   */
+  events: Array<EventData | AmqpAnnotatedMessage>;
+  /**
+   * The error that occurred when sending the associated events to the Event Hub.
+   */
+  error: Error;
+}
+
+/**
+ * Describes the options that can be provided while creating the `EventHubBufferedProducerClient`.
+ */
+export interface EventHubBufferedProducerClientOptions extends EventHubClientOptions {
+  /**
+   * The maximum number of events to buffer.
+   */
+  maxBufferedEventCount?: number;
+  /**
+   * The maximum amount of time to wait before sending a batch of messages to the Event Hub.
+   */
+  maxWaitTimeInMs?: number;
+  /**
+   * The handler to call once a batch has successfully published.
+   */
+  onSendEventsSuccessHandler?: (ctx: OnSendEventsSuccessContext) => Promise<void>;
+  /**
+   * The handler to call when a batch fails to publish.
+   */
+  onSendEventsErrorHandler: (ctx: OnSendEventsErrorContext) => Promise<void>;
+}
+
+/**
+ * Options to configure the `close` method on the `EventHubBufferedProducerClient`.
+ */
+export interface BufferedCloseOptions extends OperationOptions {
+  /**
+   * When `true`, all buffered events that are pending should be sent before closing.
+   * When `false`, abandon all buffered events and close immediately.
+   * Defaults to `false`.
+   */
+  flush?: boolean;
+}
+
+/**
+ * Options to configure the `enqueueEvents` method on the `EventHubBufferedProcuerClient`.
+ */
+export interface EnqueueEventOptions extends SendBatchOptions {}
+
+/**
+ * The `EventHubBufferedProducerClient`is used to publish events to a specific Event Hub.
+ *
+ * Depending on the options specified when events are enqueued, they may be
+ * automatically assigned to a partition, grouped according to the specified partition key,
+ * or assigned a specifically requested partition.
+ *
+ * The `EventHubBufferedProducerClient` does not publish events immediately.
+ * Instead, events are buffered so they can be efficiently batched and published
+ * when the batch is full or the `maxWaitTimeInMs` has elapsed with no new events
+ * enqueued.
```

Comment:
NIT: Since this paragraph describes the client purpose, I feel like it should be before the one preceding it.